### DISCUSSION TITLE
Multiple Thread Overwrites

### DISCUSSION THREAD

**Discussion by jgarte** [msiemens/tinydb#452] [Q&A] [answered]
Hi, can multiple threads overwrite each other's writes? 

Does tinydb handle something like this?

**Comment by msiemens** [comment, marked as answer]
@jgarte I've taken the liberty to move this to the discussions section as I think that's the best place to, well, discuss questions like these 🙂

Basically TinyDB doesn't handle this – at all. TinyDB expects you to either only write data from a single thread or use some kind of locking to make sure multiple threads don't create data corruption issues. The reasoning behind this is that TinyDB's design goals are 1) simplicity and 2) extensibility. The latter in particular means that there is no one-size-fits-all locking mechanism that would work with every imaginable storage backend that one could implement. And even if writing to the storage is atomic, TinyDB's `insert` and `update` methods are not.

Of course, one can implement a locking middleware that could work with any storage backend and would use something like use Python's [`Lock` object](https://docs.python.org/3/library/threading.html#lock-objects). Another alternative is to use something like [TinyRecord](https://github.com/eugene-eeo/tinyrecord) which tires to minimize the time spent in a lock by implementing atomic transactions.

Does that help?

**Reply by jgarte** [reply]
> @jgarte I've taken the liberty to move this to the discussions section as I think that's the best place to, well, discuss questions like these slightly_smiling_face
> 
> Basically TinyDB doesn't handle this – at all. TinyDB expects you to either only write data from a single thread or use some kind of locking to make sure multiple threads don't create data corruption issues. The reasoning behind this is that TinyDB's design goals are 1) simplicity and 2) extensibility. The latter in particular means that there is no one-size-fits-all locking mechanism that would work with every imaginable storage backend that one could implement. And even if writing to the storage is atomic, TinyDB's `insert` and `update` methods are not.
> 
> Of course, one can implement a locking middleware that could work with any storage backend and would use something like use Python's [`Lock` object](https://docs.python.org/3/library/threading.html#lock-objects). Another alternative is to use something like [TinyRecord](https://github.com/eugene-eeo/tinyrecord) which tires to minimize the time spent in a lock by implementing atomic transactions.
> 
> Does that help?

Thank you for your insightful response. This is definitely what I was looking to understand/clarify.  It is much appreciated!
